### PR TITLE
common: allow layering more than one utility provider

### PR DESCRIPTION
**Pull request by a-ilango**
This patch enables layering utility providers on top of each other. This allows providers like ofi_mrail (multi-rail) to layer over ofi_rxm provider.

- Ordering would be enforced if `hints->fabric_attr->name` lists more than two providers. The rightmost provider would be the top layer.
- Otherwise the existing behavior of not checking the order is maintained.
- The patch also allows excluding providers with a caret symbol - `^`. This is for internal use only and enables an utility provider to not layer over itself or certain other providers. Excluded providers should be listed at the end.

This PR is WIP and needs some more changes around `ofix_getinfo`.

### CODE REVIEW

File: src/fabric.c [hunk 1]
```diff
@@ -648,19 +666,19 @@ int DEFAULT_SYMVER_PRE(fi_getinfo)(uint32_t version, const char *node,
...
 	if (hints && hints->fabric_attr && hints->fabric_attr->prov_name) {
-		util_name = ofi_util_name(hints->fabric_attr->prov_name,
-					  &util_len);
-		core_name = ofi_core_name(hints->fabric_attr->prov_name,
-					  &core_len);
+		prov_vec = ofi_split_and_alloc(hints->fabric_attr->prov_name, ";");
+		if (!prov_vec)
+			return -FI_ENOMEM;
+
+		for (; prov_vec[count]; count++);
```
Comment (shefty) [inline src/fabric.c]: It would make sense to modify ofi_split_and_alloc() to return the number of non-null entries.

**Comment by a-ilango** [comment]
Updated the PR with remaining changes. It's now ready for review.

File: src/fabric.c [hunk 2]
```diff
@@ -568,59 +568,77 @@ static void ofi_set_prov_attr(struct fi_fabric_attr *attr,
...
 static int ofi_layering_ok(const struct fi_provider *provider,
-			   const char *util_name, size_t util_len,
-			   const char *core_name, size_t core_len,
+			   char **prov_vec, size_t count,
 			   uint64_t flags)
 {
+	char *prov_name;
+	int i;
+
+	/* Excluded providers are assumed to be at the end */
```
Comment (shefty) [inline src/fabric.c]: nit: I would word as MUST be at the end.

File: src/fabric.c [hunk 3]
```diff
@@ -568,59 +568,77 @@ static void ofi_set_prov_attr(struct fi_fabric_attr *attr,
...
-		     strncasecmp(core_name, provider->name, core_len)))
-			return 0;
 	}
 
-	return 1;
+	if ((count == 2) && ofi_has_util_prefix(prov_vec[0]) &&
+	    !ofi_has_util_prefix(prov_vec[1]))
+		prov_name = prov_vec[0];
+	else
+		prov_name = prov_vec[count - 1];
```
Comment (shefty) [inline src/fabric.c]: The higher-level logic for the above series of if checks doesn't clearly convey what the rules are to include/exclude a provider based on the remaining provider list.  Specifically, it's not clear why the count value should matter when determining if the layering is okay.   At the very least we probably need to add comments (yikes) here to describe the rules.

**Comment by a-ilango** [comment]
@shefty I've pulled out the string modification into utility functions. Also added code to append 
 "^<util_prov_name>" to core hints prov_name. Please take a look.

File: prov/util/src/util_attr.c [hunk 4]
```diff
@@ -132,6 +132,41 @@ const char *ofi_core_name(const char *str, size_t *len)
 
 }
 
+int ofi_exclude_prov_name(char **prov_name, const char *util_prov_name)
```
Comment (shefty) [inline prov/util/src/util_attr.c]: Please rename prov_name t prov_list or prov_name_list

Comment (a-ilango) [reply prov/util/src/util_attr.c]: Thanks! I'll rename the parameter.

**Pull request merged**